vm/state_migration/src/lib.rs

```diff
@@ -0,0 +1,259 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+//! Common code that's shared across all migration code.
+//! Each network upgrade / state migration code lives in their own module.
+
+use address::Address;
+use cid::Cid;
+use clock::ChainEpoch;
+use ipld_blockstore::BlockStore;
+use state_tree::StateTree;
+use vm::{ActorState, TokenAmount};
+
+use async_std::sync::Arc;
+use rayon::ThreadPoolBuildError;
+use std::collections::{HashMap, HashSet};
+
+pub mod nv12;
+
+pub const ACTORS_COUNT: usize = 11;
+
+pub type Migrator<BS> = Arc<dyn ActorMigration<BS> + Send + Sync>;
+pub type MigrationResult<T> = Result<T, MigrationError>;
+
+#[derive(thiserror::Error, Debug)]
+pub enum MigrationError {
+    // FIXME: use underlying concrete types when possible.
+    #[error("Failed creating job for state migration: {0}")]
+    MigrationJobCreate(String),
+    #[error("Failed running job for state migration: {0}")]
+    MigrationJobRun(String),
+    #[error("Flush failed post migration: {0}")]
+    FlushFailed(String),
+    #[error("Failed writing to blockstore: {0}")]
+    BlockStoreWrite(String),
+    #[error("Failed reading from blockstore: {0}")]
+    BlockStoreRead(String),
+    #[error("Migrator not found for cid: {0}")]
+    MigratorNotFound(Cid),
+    #[error("Failed updating new actor state: {0}")]
+    SetActorState(String),
+    #[error("State tree creation failed")]
+    StateTreeCreation(String),
+    #[error("Incomplete migration specification with {0} code CIDs")]
+    IncompleteMigrationSpec(usize),
+    #[error("Thread pool creation failed: {0}")]
+    ThreadPoolCreation(ThreadPoolBuildError),
+    #[error("Migration failed")]
+    Other,
+}
+
+pub struct StateMigration<BS> {
+    migrations: HashMap<Cid, Migrator<BS>>,
+    deferred_code_ids: HashSet<Cid>,
+}
+
+impl<BS: BlockStore + Send + Sync> StateMigration<BS> {
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Self {
+            migrations: HashMap::new(),
+            deferred_code_ids: HashSet::new(),
+        }
+    }
+
+    pub fn add_migrator(&mut self, prior_cid: Cid, migrator: Migrator<BS>) {
+        self.migrations.insert(prior_cid, migrator);
+    }
+
+    pub fn migrate_state_tree(
+        &mut self,
+        store: Arc<BS>,
+        prior_epoch: ChainEpoch,
+        actors_in: StateTree<BS>,
+        mut actors_out: StateTree<BS>,
+    ) -> MigrationResult<Cid> {
+        if self.migrations.len() + self.deferred_code_ids.len() != ACTORS_COUNT {
+            return Err(MigrationError::IncompleteMigrationSpec(
+                self.migrations.len(),
+            ));
+        }
+
+        let cpus = num_cpus::get();
+        let chan_size = cpus / 2;
```

Comment:
I want to make the case that `chan_size` should be set to `cpus` but I don't know why you chose `cpus / 2`, could you add a comment?

Reply:
I think this was empirically calculated based on Rahuls cloud machine...

Reply:
As i was testing the migration run-times, i noticed that increasing channel size made the migration time jump by 12-13 secs. 
For example (from my raw test logs):
```
16 threads 
16 chan size
16 outchan - run-time 80 secs
```
OTOH decreasing the channel size to less than half the cpu core count increased the run-time too. I am not exactly sure why it increases in the first case. We can file an issue to tune it in a later PR.